### PR TITLE
HSEARCH-2624 Add a "tenantId" parameter

### PR DESCRIPTION
**Pull request by mincong-h**
Please do not merge this PR.

I'm currently blocked with the code configuration. When implementing the multi-tenancy methods, the following method has always been skipped automatically:

    MassIndexingJobWithMultiTenancyTest#shouldHandleTenantIds

I spent many time on it and still cannot find why... Please help.

### CODE REVIEW

**Comment by yrodiere** [comment]
@mincong-h I added a commit that fixes the main issue, i.e. you didn't add all the necessary dependencies, and also some other unrelated problems. Feel free to squash my commit into yours, I only kept it separated so you can see what changed.
Note that the test still fails, but that's another problem...

**Comment by mincong-h** [comment]
Thanks @yrodiere. Yes, I'm aware that there're other problem(s). I'll continue work on it and leave you a comment once I finish the PR.

**Comment by mincong-h** [comment]
Hi Yoann, here's the updated PR! I didn't have enough time to polish the code before the end of the evening, especially the integration test. I was struggled with the missing `log4j.properties` during several hours. But the code here should be able to handle the multi-tenancy correctly, and it's ready to be reviewed. To simplify the task, let's squash all the commits into one before the rebase-and-merge.

**Comment by mincong-h** [comment]
I'll handle the conflict this week

**Comment by yrodiere** [comment]
> I'll handle the conflict this week

Ok, thanks. I'm afraid it's all because of my other PR... Note that I removed `MassIndexerUtil` in my other PR (moved the code to `SerializationUtil`), but feel free to recreate it with only your new code in it.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/MassIndexerUtil.java [hunk 1]
```diff
@@ -54,4 +59,52 @@ public static String serializeCriteria(Set<Criterion> criteria)
...
+	 *
+	 * @param entityManagerFactory entity manager factory
+	 * @param tenantId tenant ID, can be {@literal null} or empty.
+	 *
+	 * @return a new session
+	 */
+	public static Session openSession(EntityManagerFactory entityManagerFactory, String tenantId) {
+		SessionFactory sessionFactory = entityManagerFactory.unwrap( SessionFactory.class );
+		Session session;
+		if ( tenantId == null || tenantId.isEmpty() ) {
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/MassIndexerUtil.java]: Just so you know, you can also simply use `org.hibernate.search.util.StringHelper.isEmpty(String)`.

File: jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobParametersBuilderTest.java [hunk 2]
```diff
@@ -115,4 +120,25 @@ public void testRestrictedBy_twoRestrictionTypes() {
...
 				.restrictedBy( Restrictions.isEmpty( "dummy" ) );
 	}
+
+	@Test
+	public void testTenantId_null() {
+		try {
+			MassIndexingJob.parameters().forEntity( String.class ).tenantId( null );
+		}
+		catch (NullPointerException e) {
+			assertEquals( e.getMessage(), "Your tenantId is null, please provide a valid tenant ID.", e.getMessage() );
```
Comment (yrodiere) [inline jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobParametersBuilderTest.java]: This test will pass if the method doesn't throw an exception, which is not what you want. When checking whether code throws an exception, you'll probably find it easier to use an `ExpectedException` rule. You can have a look at `org.hibernate.search.test.util.impl.LoggingCreationTest` to see how to use it.

File: jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobWithMultiTenancyIT.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import org.fest.util.Collections;
+
+import static org.fest.assertions.Assertions.assertThat;
+
+/**
+ * @author Mincong Huang
+ */
+public class MassIndexingJobWithMultiTenancyIT extends SearchTestBase {
+
+	public static final Dialect DIALECT = new H2Dialect();
```
Comment (yrodiere) [inline jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobWithMultiTenancyIT.java]: This seems unnecessary.

**Review by yrodiere — COMMENTED** [review]
Looks great, thanks! I'll let you resolve the conflicts and I'll merge it after that.

**Comment by mincong-h** [comment]
Hi Yoann, I resolved all the conflicts, but I didn't have enough time to polish the code. I'll continue tomorrow morning. Please wait a little bit for the review.

File: jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobWithMultiTenancyIT.java (hunk 3 above)
Comment (mincong-h) [reply jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobWithMultiTenancyIT.java]: Yes, you're right. Actually, I deleted it once and then job failed, but it must be due to another cause which I didn't find out. It has been removed now.

File: jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobParametersBuilderTest.java (hunk 2 above)
Comment (mincong-h) [reply jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobParametersBuilderTest.java]: Thanks!

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/MassIndexerUtil.java (hunk 1 above)
Comment (mincong-h) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/MassIndexerUtil.java]: :+1:

**Comment by mincong-h** [comment]
I finished my modification, please review again @yrodiere! :grin:

**Comment by yrodiere** [comment]
Nice! Merged, thanks :)

**Comment by mincong-h** [comment]
You're welcome. Have a nice day!

**Pull request merged**